src/main/credentialServer/httpCredentialHelperServer.ts

```diff
@@ -0,0 +1,316 @@
+    if (commandName) {
+      return this.dispatchTable[method]?.[commandName];
+    }
+
+    return undefined;
+  }
+
+  async get(helperName: string, data: string, request: http.IncomingMessage, response: http.ServerResponse): Promise<void> {
+    return await this.doNamedCommand(requireJSONOutput, helperName, 'get', data, request, response);
+  }
+
+  async list(helperName: string, data: string, request: http.IncomingMessage, response: http.ServerResponse): Promise<void> {
+    return await this.doNamedCommand(requireJSONOutput, helperName, 'list', data, request, response);
+  }
+
+  async erase(helperName: string, data: string, request: http.IncomingMessage, response: http.ServerResponse): Promise<void> {
+    return await this.doNamedCommand(requireNoOutput, helperName, 'erase', data, request, response);
+  }
+
+  async store(helperName: string, data: string, request: http.IncomingMessage, response: http.ServerResponse): Promise<void> {
+    return await this.doNamedCommand(requireNoOutput, helperName, 'store', data, request, response);
+  }
+
+  protected async doNamedCommand(outputChecker: checkFunctionOutputType,
+    helperName: string,
+    commandName: string,
+    data: string,
+    request: http.IncomingMessage,
+    response: http.ServerResponse): Promise<void> {
+    let stderr: string;
+
+    try {
+      const stdout = await this.runWithInput(data, helperName, [commandName]);
+
+      if (outputChecker(stdout)) {
+        response.writeHead(200, { 'Content-Type': 'text/plain' });
+        response.write(stdout);
+
+        return await Promise.resolve();
+      }
+      stderr = stdout;
+    } catch (err: any) {
+      stderr = err.stderr ?? err.stdout ?? '';
+    }
+    console.debug(`credentialServer: ${ commandName }: writing back status 400, error: ${ stderr }`);
+    response.writeHead(400, { 'Content-Type': 'text/plain' });
+    response.write(stderr);
+
+    return await Promise.resolve();
+  }
+
+  /**
+   * Returns the name of the credential-helper to use (which is a suffix of the helper `docker-credential-`).
+   *
+   * Note that callers are responsible for catching exceptions, which usually happen if the
+   * `$HOME/docker/config.json` doesn't exist, it's JSON is corrupt, or it doesn't have a `credsStore` field.
+   */
+  protected async getCredentialHelperName(): Promise<string> {
+    const home = findHomeDir();
+    const dockerConfig = path.join(home ?? '', '.docker', 'config.json');
+    const contents = JSON.parse((await fs.promises.readFile(dockerConfig, { encoding: 'utf-8' })).toString());
+    const credStore = contents['credsStore'];
+
+    if (!credStore) {
+      throw new Error(`No credStore field in ${ dockerConfig }`);
+    }
+
+    return credStore;
+  }
+
+  closeServer() {
+    for (const server of this.servers) {
+      server.close();
+    }
+  }
+
+  protected runWithInput(data: string, command: string, args: string[]): Promise<string> {
```

Comment:
We can get away with less code if we make `spawnFile` take a `Readable` as `stdin`:
```diff
diff --git src/utils/childProcess.ts src/utils/childProcess.ts
index cf0de6d2..bebafec4 100644
--- src/utils/childProcess.ts
+++ src/utils/childProcess.ts
@@ -162,7 +162,7 @@ export async function spawnFile(
     (typeof options.encoding === 'string') ? options.encoding : options.encoding?.stdout,
     (typeof options.encoding === 'string') ? options.encoding : options.encoding?.stderr,
   ];
-  const stdStreams: [undefined, stream.Writable | undefined, stream.Writable | undefined] = [undefined, undefined, undefined];
+  const stdStreams: [undefined | stream.Readable, stream.Writable | undefined, stream.Writable | undefined] = [undefined, undefined, undefined];
   let mungedStdio: StdioOptions = 'pipe';

   // If we're piping to a stream, and we need to override the encoding, then
@@ -175,6 +175,9 @@ export async function spawnFile(

       if (isLog(original)) {
         munged = await original.fdStream;
+      } else if (i == 0 && original instanceof stream.Readable) {
+        munged = 'pipe';
+        stdStreams[i] = original;
       } else {
         munged = original;
       }
@@ -201,7 +204,10 @@ export async function spawnFile(
   const result: { stdout?: string, stderr?: string } = {};

   if (Array.isArray(mungedStdio)) {
-    for (const i of [1, 2]) {
+    if (stdStreams[0] instanceof stream.Readable && child.stdin) {
+      stdStreams[0].pipe(child.stdin);
+    }
+    for (const i of [1, 2] as const) {
       if (mungedStdio[i] === 'pipe') {
         const encoding = encodings[i];
         const childStream = child[resultMap[i]];
```
This means we can construct the input easier:
```ts
 protected async runWithInput(data: string, command: string, args: string[]): Promise<string> {
   const body = stream.Readable.from(data);
   const { stdout } = await spawnFile(command, args, {
     stdio: [body, 'pipe', 'pipe'],
     windowsHide: true,
   });
   return stdout;
 }
```
Note that this means on exception `stderr` is `''` instead of `undefined`, so we'd need `err.stderr || err.stdout || ''` instead of `??`.